pkg/controllers/node/termination/controller.go

```diff
@@ -157,6 +164,34 @@ func (c *Controller) removeFinalizer(ctx context.Context, n *v1.Node) error {
 	return nil
 }
 
+func (c *Controller) taintExpiredNode(ctx context.Context, node *v1.Node) (*time.Time, error) {
+	nodeGracePeriodExpirationTime, err := c.nodeExpirationTime(node)
+	if err != nil {
+		return nil, err
+	} else if nodeGracePeriodExpirationTime != nil && time.Now().After(*nodeGracePeriodExpirationTime) {
+		taint := v1beta1.DisruptionNonGracefulShutdown
+		logging.FromContext(ctx).With("taint.Key", taint.Key).With("taint.Effect", taint.Effect).With("taint.Value", taint.Value).Infof("terminationGracePeriod for the node has expired, tainting")
+		if err := c.terminator.Taint(ctx, node, taint); err != nil {
+			return nil, fmt.Errorf("tainting node: %w", err)
+		}
+	}
+	return nodeGracePeriodExpirationTime, nil
+}
+
+func (c *Controller) nodeExpirationTime(node *v1.Node) (*time.Time, error) {
+	if expirationTimeString, exists := node.ObjectMeta.Annotations[v1beta1.NodeExpirationTimeAnnotationKey]; exists {
+		c.recorder.Publish(terminatorevents.NodeTerminationGracePeriod(node, expirationTimeString))
```

Comment:
Should we just publish this event once when the annotation is applied?

Reply:
Yep agree, time to learn how to plumb the recorder through in the nodeclaim controller.